File: page-storage-system/storage/page_manager.py
import json
import os
from pathlib import Path
from typing import Union

from .constants import PAGE_SIZE
from .errors import StorageError
from .file_manager import FileManager


class PageManager:
    """Owns page allocation state and validates all page access."""

    def __init__(self, root: Union[str, Path]):
        self.root = Path(root)
        self.root.mkdir(parents=True, exist_ok=True)
        self.file_manager = FileManager(self.root / "pages.dat")
        self.meta_path = self.root / "page_allocation.json"
        self._allocated = set()
        self._free = []
        self._load_or_initialize()

    def _load_or_initialize(self):
        if not self.meta_path.exists():
            self._persist()
            return
        try:
            state = json.loads(self.meta_path.read_text(encoding="utf-8"))
            self._allocated = {int(v) for v in state.get("allocated", [])}
            self._free = sorted({int(v) for v in state.get("free", [])})
        except (OSError, ValueError, TypeError, json.JSONDecodeError) as exc:
            raise StorageError("PAGE_METADATA_CORRUPT", f"页分配元数据损坏: {exc}") from exc

    def _persist(self):
        state = {"allocated": sorted(self._allocated), "free": sorted(self._free)}
        tmp = self.meta_path.with_suffix(".tmp")
        try:
            tmp.write_text(json.dumps(state, ensure_ascii=False, indent=2), encoding="utf-8")
            os.replace(tmp, self.meta_path)
        except OSError as exc:
            raise StorageError("FILE_IO_ERROR", f"持久化页分配状态失败: {exc}") from exc

    def is_allocated(self, page_id: int) -> bool:
        return page_id in self._allocated

    def _require_allocated(self, page_id: int):
        if not isinstance(page_id, int) or isinstance(page_id, bool) or page_id < 0:
            raise StorageError("INVALID_PAGE_ID", "pageId 必须是非负整数", page_id=page_id)
        if page_id not in self._allocated:
            raise StorageError("PAGE_NOT_ALLOCATED", "页不存在或已经释放", page_id=page_id)

    def allocate_page(self) -> int:
        if self._free:
            page_id = self._free.pop(0)
            self.file_manager.write_at(page_id, bytes(PAGE_SIZE))
        else:
            page_id = self.file_manager.append_zero_page()
        self._allocated.add(page_id)
        try:
            self._persist()
        except Exception:
            self._allocated.discard(page_id)
            if page_id not in self._free:
                self._free.append(page_id)
                self._free.sort()
            raise
        return page_id

    def free_page(self, page_id: int) -> None:
        self._require_allocated(page_id)
        self._allocated.remove(page_id)
        self._free.append(page_id)
        self._free.sort()
        try:
            self._persist()
        except Exception:
            self._free.remove(page_id)
            self._allocated.add(page_id)
            raise

    def read_page(self, page_id: int) -> bytes:
        self._require_allocated(page_id)
        return self.file_manager.read_at(page_id)

    def write_page(self, page_id: int, data: bytes) -> int:
        self._require_allocated(page_id)
        if len(data) != PAGE_SIZE:
            raise StorageError("INVALID_PAGE_SIZE", "写入页必须恰好为 4096 字节", page_id=page_id)
        return self.file_manager.write_at(page_id, data)

    def sync(self):
        self.file_manager.sync()

    @property
    def allocated_count(self) -> int:
        return len(self._allocated)

File: page-storage-system/src/main/java/org/chainpage/storage/PageManager.java
package org.chainpage.storage;

import com.fasterxml.jackson.databind.JsonNode;
import java.io.IOException;
import java.nio.file.*;
import java.util.*;

public final class PageManager {
    private final Path meta; private final FileManager file;
    private final SortedSet<Integer> allocated=new TreeSet<>(), free=new TreeSet<>();
    private final Map<Integer,Long> generations=new HashMap<>();
    private Runnable invalidator=()->{}; private java.util.function.IntConsumer pageInvalidator=id->{};
    PageManager(Path root){file=new FileManager(root.resolve("pages.dat"));meta=root.resolve("page_allocation.json");load();}
    synchronized void setPageInvalidator(java.util.function.IntConsumer c){pageInvalidator=c;}
    synchronized void load(){allocated.clear();free.clear();generations.clear();
        if(!Files.exists(meta)){persist();return;}
        try{
            JsonNode n=JsonFiles.parse(Files.readAllBytes(meta),"PAGE_METADATA_CORRUPT");
            if(!n.isObject()||!n.has("allocated")||!n.has("free"))throw corrupt("字段缺失");
            readIds(n.get("allocated"),allocated);readIds(n.get("free"),free);
            if(!Collections.disjoint(allocated,free))throw corrupt("allocated/free 重叠");
            JsonNode gs=n.get("generations");if(gs!=null){if(!gs.isObject())throw corrupt("generations 非对象");gs.fields().forEachRemaining(e->{try{int id=Integer.parseInt(e.getKey());JsonNode value=e.getValue();if(!value.isIntegralNumber()||!value.canConvertToLong())throw new Exception();long g=value.longValue();if(id<0||g<0)throw new Exception();generations.put(id,g);}catch(Exception x){throw corrupt("generation 非法");}});}
            long size=Files.size(file.path()); long full=size/FileManager.PAGE_SIZE;if(size%FileManager.PAGE_SIZE!=0)file.truncate(full*FileManager.PAGE_SIZE);
            int pages=Math.toIntExact(full);if(allocated.stream().anyMatch(i->i>=pages)||free.stream().anyMatch(i->i>=pages))throw corrupt("页号超出数据文件");
            boolean changed=false;for(int i=0;i<pages;i++)if(!allocated.contains(i)&&!free.contains(i)){free.add(i);changed=true;}if(changed)persist();
        }catch(IOException|ArithmeticException e){throw new StorageException("PAGE_METADATA_CORRUPT",e.getMessage(),null,e);}
    }
    private void readIds(JsonNode n,Set<Integer> out){if(!n.isArray())throw corrupt("分配表非数组");for(JsonNode v:n){if(!v.isIntegralNumber()||!v.canConvertToInt()||v.intValue()<0||!out.add(v.intValue()))throw corrupt("非法或重复页号");}}
    private StorageException corrupt(String m){return new StorageException("PAGE_METADATA_CORRUPT",m);}
    synchronized void persist(){Map<String,Object> m=JsonFiles.map();m.put("allocated",allocated);m.put("free",free);Map<String,Long> gs=new TreeMap<>();generations.forEach((k,v)->gs.put(k.toString(),v));m.put("generations",gs);JsonFiles.replace(meta,JsonFiles.json(m));}
    public synchronized int allocatePage(){boolean reuse=!free.isEmpty();int id=reuse?free.first():(int)file.pageCount();long old=generation(id);
        if(reuse)file.writeAt(id,new byte[FileManager.PAGE_SIZE]);else{int made=file.appendZero();if(made!=id)throw new StorageException("FILE_IO_ERROR","并发扩页冲突");}
        file.sync();if(reuse)free.remove(id);allocated.add(id);generations.put(id,old+1);
        try{persist();CrashHooks.hit("page_allocate");return id;}catch(RuntimeException e){allocated.remove(id);generations.put(id,old);if(reuse)free.add(id);else file.truncate((long)id*FileManager.PAGE_SIZE);throw e;}}
    public synchronized void freePage(int id){requireAllocated(id);pageInvalidator.accept(id);allocated.remove(id);free.add(id);try{persist();CrashHooks.hit("page_free");}catch(RuntimeException e){free.remove(id);allocated.add(id);throw e;}}
    public synchronized byte[] readPage(int id){requireAllocated(id);return file.readAt(id);}
    public synchronized void writePage(int id,byte[] data){requireAllocated(id);FileManager.page(data,id);file.writeAt(id,data);}
    public synchronized void sync(){file.sync();}
    public synchronized void requireAllocated(int id){FileManager.valid(id);if(!allocated.contains(id))throw new StorageException("PAGE_NOT_ALLOCATED","页不存在或已经释放",id);}
    public synchronized boolean isAllocated(int id){return id>=0&&allocated.contains(id);}
    public synchronized long generation(int id){return generations.getOrDefault(id,0L);}
    public synchronized int allocatedCount(){return allocated.size();}
    synchronized Set<Integer> allocatedIds(){return new TreeSet<>(allocated);}
    Path dataPath(){return file.path();}
}
